### DISCUSSION TITLE
How to use the matplotlib in the project and it can work to run in the android

### DISCUSSION THREAD

**Discussion by ccinsz** [beeware/briefcase#1289] [General]
Hi, I use the newest version of the matplotlib-3.7.1, and it raises AttributeError: module 'os' has no attribute 'add_dll_directory'. How can I use the matplotlib successfully in my project?

And when I use toga-android~=0.3.1, it also raises ValueError("Unknown argument type", arg, type(arg)), so I change to use toga-android==0.3.0dev39, and it raises AttributeError: module 'os' has no attribute 'add_dll_directory'.

**Comment by freakboy3742** [comment]
It's not possible to use arbitrary versions of Python modules that have binary extensions. In this case, matplotlib 3.6.0 and 3.1.2 are the only versions that have any support on Android; you can check https://chaquo.com/pypi-7.0/matplotlib/ and https://chaquo.com/pypi-13.1/matplotlib/ to see the versions that have binary wheels available.

The error `add_dll_directory` error you're seeing is caused because Python is not finding a binary wheel for matplotlib 3.7.1, and is finding an alternate wheel (or compiling one), but that wheel is a *windows* wheel, not an Android. This *shouldn't* be succeeding; however, without logs of the build, it's impossible to say what has happened.

As for the error with Toga - it's impossible to say what is going on there, because you've provided no context for how you're generating the error.

**Comment by ccinsz** [comment]
Yes, I also try matplotlib 3.6.0 in Python 3.8 or Python 3.10, they have the same problem. I use "briefcase run android -u", and it raises
04-14 03:40:32.670 13348 13348 E Python  : Please note and check the following:
04-14 03:40:32.670 13348 13348 E Python  :
04-14 03:40:32.670 13348 13348 E Python  :   * The Python version is: Python3.8 from ""
04-14 03:40:32.670 13348 13348 E Python  :   * The NumPy version is: "1.24.3"
04-14 03:40:32.670 13348 13348 E Python  :
04-14 03:40:32.670 13348 13348 E Python  : and make sure that they are the versions you expect.
04-14 03:40:32.670 13348 13348 E Python  : Please carefully study the documentation linked above for further help.
04-14 03:40:32.670 13348 13348 E Python  :
04-14 03:40:32.670 13348 13348 E Python  : Original error was: No module named 'numpy.core._multiarray_umath'
04-14 03:40:32.670 13348 13348 E Python  :
04-14 03:40:32.671 13348 13348 D Python  : Native invocation done.

**Comment by ccinsz** [comment]
And I try to change the version of the numpy, but it can't work.

**Reply by freakboy3742** [reply]
It's very hard to debug a problem when you only provide partial information. If you want help diagnosing a problem, we need *full* reproduction instructions, and *full* logs of any crashes - not just a "I tried matplotlib and here's the last 10 lines of a stack trace".

**Reply by ccinsz** [reply]
The full logs like that:

===========================================================================
--------- beginning of main
04-14 04:07:36.689  5760  5760 D MainActivity: onCreate() start
04-14 04:07:36.689  5760  5760 D MainActivity: onCreate(): captured stdout/stderr
04-14 04:07:36.689  5760  5792 D stdio   : Starting to capture stdout/stderr to Android log.
04-14 04:07:36.689  5760  5792 D stdio   : stdout now successfully routes into the stdio logger.
04-14 04:07:36.732  5760  5760 D MainActivity: Python stdlib already exists for
stdlib/pythonhome.1e99b1b3d8cb4a8ad361e37b0dca59664cec0560ec60d026b8a3c64331f34565.arm64-v8a.zip
04-14 04:07:36.732  5760  5760 D MainActivity: Unpacking Python assets to
/data/user/0/com.example.helloworld/files/python/user_code
04-14 04:07:39.400  5760  5760 D MainActivity: Replacing old lastUpdateTime = 1681414790387 with actualLastUpdateTime =
1681416456143
04-14 04:07:39.400  5760  5760 D MainActivity: unpackPython() complete
04-14 04:07:39.400  5760  5760 D MainActivity: setPythonEnvVars() start
04-14 04:07:39.400  5760  5760 V MainActivity: pythonHome=/data/user/0/com.example.helloworld/files/python/stdlib
04-14 04:07:39.400  5760  5760 D MainActivity: setPythonEnvVars() complete
04-14 04:07:39.400  5760  5760 D MainActivity: Computing Python version.
04-14 04:07:39.400  5760  5760 D MainActivity: Computed Python version: 3.8
04-14 04:07:39.402  5760  5760 I Python  : Start Python runtime...
04-14 04:07:39.402  5760  5760 D Python  : PYTHONHOME=/data/user/0/com.example.helloworld/files/python/stdlib
04-14 04:07:39.403  5760  5760 D Python  :
PYTHONPATH=/data/user/0/com.example.helloworld/files/python/stdlib/lib/python3.8/:/data/user/0/com.example.helloworld/fi
les/python/user_code/app_packages:/data/user/0/com.example.helloworld/files/python/user_code/app
04-14 04:07:39.403  5760  5760 D Python  : Not setting RUBICON_LIBRARY
04-14 04:07:39.403  5760  5760 I Python  : Adding Android logging module to default modules...
04-14 04:07:39.403  5760  5760 D Python  : Initializing Python runtime...
04-14 04:07:39.415  5760  5760 D Python  : Replacing sys.stdout/sys.stderr with Android log wrappers...
04-14 04:07:39.416  5760  5760 I Python  : sys.stdout/stderr replaced with Android log wrappers.
04-14 04:07:39.416  5760  5760 V Python  : Import rubicon...
04-14 04:07:39.444  5760  5760 V Python  : Got rubicon python module
04-14 04:07:39.444  5760  5760 V Python  : Got method dispatch handler
04-14 04:07:39.445  5760  5760 I Python  : Python runtime started.
04-14 04:07:39.445  5760  5760 D MainActivity: Python.init() complete
04-14 04:07:39.445  5760  5760 D MainActivity: Python.run() start
04-14 04:07:39.445  5760  5760 D Python  : Running 'helloworld' as __main__...
04-14 04:07:39.629  5760  5760 E Python  : Application quit abnormally!
04-14 04:07:39.629  5760  5760 E Python  : Traceback (most recent call last):
04-14 04:07:39.629  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/user_code/app_packages/numpy/core/__init__.py", line 23, in <module>
04-14 04:07:39.630  5760  5760 E Python  :     from . import multiarray
04-14 04:07:39.630  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/user_code/app_packages/numpy/core/multiarray.py", line 10, in <module>
04-14 04:07:39.630  5760  5760 E Python  :     from . import overrides
04-14 04:07:39.630  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/user_code/app_packages/numpy/core/overrides.py", line 6, in <module>
04-14 04:07:39.630  5760  5760 E Python  :     from numpy.core._multiarray_umath import (
04-14 04:07:39.630  5760  5760 E Python  : ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'
04-14 04:07:39.630  5760  5760 E Python  :
04-14 04:07:39.630  5760  5760 E Python  : During handling of the above exception, another exception occurred:
04-14 04:07:39.630  5760  5760 E Python  :
04-14 04:07:39.630  5760  5760 E Python  : Traceback (most recent call last):
04-14 04:07:39.630  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/stdlib/lib/python3.8/runpy.py", line 194, in _run_module_as_main
04-14 04:07:39.630  5760  5760 E Python  :     return _run_code(code, main_globals, None,
04-14 04:07:39.630  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/stdlib/lib/python3.8/runpy.py", line 87, in _run_code
04-14 04:07:39.630  5760  5760 E Python  :     exec(code, run_globals)
04-14 04:07:39.630  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/user_code/app/helloworld/__main__.py", line 1, in <module>
04-14 04:07:39.630  5760  5760 E Python  :     from helloworld.app import main
04-14 04:07:39.630  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/user_code/app/helloworld/app.py", line 7, in <module>
04-14 04:07:39.630  5760  5760 E Python  :     import matplotlib
04-14 04:07:39.630  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/user_code/app_packages/matplotlib/__init__.py", line 108, in <module>
04-14 04:07:39.630  5760  5760 E Python  :     import numpy
04-14 04:07:39.630  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/user_code/app_packages/numpy/__init__.py", line 141, in <module>
04-14 04:07:39.631  5760  5760 E Python  :     from . import core
04-14 04:07:39.631  5760  5760 E Python  :   File
"/data/user/0/com.example.helloworld/files/python/user_code/app_packages/numpy/core/__init__.py", line 49, in <module>
04-14 04:07:39.631  5760  5760 E Python  :     raise ImportError(msg)
04-14 04:07:39.631  5760  5760 E Python  : ImportError:
04-14 04:07:39.631  5760  5760 E Python  :
04-14 04:07:39.631  5760  5760 E Python  : IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!
04-14 04:07:39.631  5760  5760 E Python  :
04-14 04:07:39.631  5760  5760 E Python  : Importing the numpy C-extensions failed. This error can happen for
04-14 04:07:39.631  5760  5760 E Python  : many reasons, often due to issues with your setup or how NumPy was
04-14 04:07:39.631  5760  5760 E Python  : installed.
04-14 04:07:39.631  5760  5760 E Python  :
04-14 04:07:39.631  5760  5760 E Python  : We have compiled some common reasons and troubleshooting tips at:
04-14 04:07:39.631  5760  5760 E Python  :
04-14 04:07:39.631  5760  5760 E Python  :     https://numpy.org/devdocs/user/troubleshooting-importerror.html
04-14 04:07:39.631  5760  5760 E Python  :
04-14 04:07:39.631  5760  5760 E Python  : Please note and check the following:
04-14 04:07:39.631  5760  5760 E Python  :
04-14 04:07:39.631  5760  5760 E Python  :   * The Python version is: Python3.8 from ""
04-14 04:07:39.631  5760  5760 E Python  :   * The NumPy version is: "1.24.3"
04-14 04:07:39.631  5760  5760 E Python  :
04-14 04:07:39.631  5760  5760 E Python  : and make sure that they are the versions you expect.
04-14 04:07:39.631  5760  5760 E Python  : Please carefully study the documentation linked above for further help.
04-14 04:07:39.631  5760  5760 E Python  :
04-14 04:07:39.631  5760  5760 E Python  : Original error was: No module named 'numpy.core._multiarray_umath'
04-14 04:07:39.631  5760  5760 E Python  :
04-14 04:07:39.631  5760  5760 D MainActivity: Python.run() end
04-14 04:07:39.631  5760  5760 D MainActivity: startPython() end
04-14 04:07:39.631  5760  5760 D MainActivity: user code onCreate() start

**Reply by ccinsz** [reply]
pyproject.toml add "matplotlib==3.6.0", the uses  "briefcase update android -d" and "briefcase run android -u" to run the project.

**Reply by mhsmith** [reply]
Based on this log, you're using an old version of the Briefcase Android template. Upgrade to the current version of Briefcase, then run `briefcase create android`.

**Reply by ccinsz** [reply]
I use the newest version of the Briefcase, but it has problem. When it run "briefcase create android", it will raises "You did not accept the Android SDK licenses. Please re-run the briefcase command and accept the Android SDK license when prompted. You may need an Internetconnection."

I use the briefcase==0.3.8 and it works.

**Reply by mhsmith** [reply]
If you're using the newest version of Briefcase, then the last thing it displays after a failure should be the name of a log file. Please attach that file to this discussion.

**Reply by ccinsz** [reply]
This is the log:
```
OS Release:      Windows 10
OS Version:      10.0.22000
Architecture:    AMD64
Platform:        Windows-10-10.0.22000-SP0

Python exe:      D:\beeware-venv\Scripts\python.exe
Python version:  3.8.7 (tags/v3.8.7:6503f05, Dec 21 2020, 17:59:51) [MSC v.1928 64 bit (AMD64)]
Virtual env:     True
Conda env:       False

Briefcase:       0.3.14
Target platform: android
Target format:   gradle

Briefcase Log:
[16:36:08]                                                                                                                                                        android_sdk.py:372
           The Android tools provided by Google have license terms that you must accept                                                                           android_sdk.py:372
           before you may use those tools.                                                                                                                        android_sdk.py:372
                                                                                                                                                                   subprocess.py:675
           >>> Running Command:                                                                                                                                    subprocess.py:676
           >>>     'C:\Users\ccinsz\AppData\Local\BeeWare\briefcase\Cache\tools\android_sdk\cmdline-tools\latest\bin\sdkmanager.bat' --licenses              subprocess.py:677
           >>> Working Directory:                                                                                                                                  subprocess.py:684
           >>>     D:\demo0524                                                                                       subprocess.py:685
           >>> Environment Overrides:                                                                                                                              subprocess.py:694
           >>>     ANDROID_SDK_ROOT=C:\Users\ccinsz\AppData\Local\BeeWare\briefcase\Cache\tools\android_sdk                                                  subprocess.py:696
           >>>     JAVA_HOME=C:\Users\ccinsz\AppData\Local\BeeWare\briefcase\Cache\tools\java                                                                subprocess.py:696
[16:36:10] >>> Return code: 0                                                                                                                                      subprocess.py:712
                                                                                                                                                                      __main__.py:40
           You did not accept the Android SDK licenses. Please re-run the briefcase command                                                                           __main__.py:41
           and accept the Android SDK license when prompted. You may need an Internet                                                                                 __main__.py:41
           connection.                                                                                                                                                __main__.py:41

Main thread traceback:
╭─────────────────────────────────────────────────────────────────────── Traceback (most recent call last) ────────────────────────────────────────────────────────────────────────╮
│ D:\beeware-venv\lib\site-packages\briefcase\__main__.py:25 in main                                                                 │
│                                                                                                                                                                                  │
│   22 │   │   command = Command(logger=logger, console=console)                                                                                                                   │
│   23 │   │   options = command.parse_options(extra=extra_cmdline)                                                                                                                │
│   24 │   │   command.parse_config(Path.cwd() / "pyproject.toml")                                                                                                                 │
│ ❱ 25 │   │   command(**options)                                                                                                                                                  │
│   26 │   except HelpText as e:                                                                                                                                                   │
│   27 │   │   logger.info()                                                                                                                                                       │
│   28 │   │   logger.info(str(e))                                                                                                                                                 │
│                                                                                                                                                                                  │
│ ╭─────────────────────────────────────────────────────────────────────────────────── locals ───────────────────────────────────────────────────────────────────────────────────╮ │
│ │       command = <briefcase.platforms.android.gradle.GradleCreateCommand object at 0x0000024864293CA0>                                                                        │ │
│ │       Command = <class 'briefcase.platforms.android.gradle.GradleCreateCommand'>                                                                                             │ │
│ │       console = <briefcase.console.Console object at 0x000002486429E100>                                                                                                     │ │
│ │             e = BriefcaseCommandError('You did not accept the Android SDK licenses. Please re-run the briefcase command\nand accept the Android SDK license when prompted.   │ │
│ │                 You may need an Internet\nconnection.\n')                                                                                                                    │ │
│ │ extra_cmdline = []                                                                                                                                                           │ │
│ │        logger = <briefcase.console.Log object at 0x00000248642ABE50>                                                                                                         │ │
│ │       options = {}                                                                                                                                                           │ │
│ │        result = 200                                                                                                                                                          │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                                                                                                  │
│ D:\beeware-venv\lib\site-packages\briefcase\commands\create.py:798 in __call__                                                     │
│                                                                                                                                                                                  │
│   795 │   def __call__(self, app: Optional[BaseConfig] = None, **options):                                                                                                       │
│   796 │   │   # Confirm host compatibility, that all required tools are available,                                                                                               │
│   797 │   │   # and that the app configuration is finalized.                                                                                                                     │
│ ❱ 798 │   │   self.finalize(app)                                                                                                                                                 │
│   799 │   │                                                                                                                                                                      │
│   800 │   │   if app:                                                                                                                                                            │
│   801 │   │   │   state = self.create_app(app, **options)                                                                                                                        │
│                                                                                                                                                                                  │
│ ╭──────────────────────────────────────────── locals ─────────────────────────────────────────────╮                                                                              │
│ │     app = None                                                                                  │                                                                              │
│ │ options = {}                                                                                    │                                                                              │
│ │    self = <briefcase.platforms.android.gradle.GradleCreateCommand object at 0x0000024864293CA0> │                                                                              │
│ ╰─────────────────────────────────────────────────────────────────────────────────────────────────╯                                                                              │
│                                                                                                                                                                                  │
│ D:\beeware-venv\lib\site-packages\briefcase\commands\base.py:520 in finalize                                                       │
│                                                                                                                                                                                  │
│   517 │   │   │   to finalize. By default, all apps will be finalized.                                                                                                           │
│   518 │   │   """                                                                                                                                                                │
│   519 │   │   self.verify_host()                                                                                                                                                 │
│ ❱ 520 │   │   self.verify_tools()                                                                                                                                                │
│   521 │   │                                                                                                                                                                      │
│   522 │   │   if app is None:                                                                                                                                                    │
│   523 │   │   │   for app in self.apps.values():                                                                                                                                 │
│                                                                                                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮                                                                                 │
│ │  app = None                                                                                  │                                                                                 │
│ │ self = <briefcase.platforms.android.gradle.GradleCreateCommand object at 0x0000024864293CA0> │                                                                                 │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯                                                                                 │
│                                                                                                                                                                                  │
│ D:\beeware-venv\lib\site-packages\briefcase\platforms\android\gradle.py:123 in verify_tools                                        │
│                                                                                                                                                                                  │
│   120 │   │   """Verify that the Android APK tools in `briefcase` will operate on this system,                                                                                   │
│   121 │   │   downloading tools as needed."""                                                                                                                                    │
│   122 │   │   super().verify_tools()                                                                                                                                             │
│ ❱ 123 │   │   AndroidSDK.verify(tools=self.tools)                                                                                                                                │
│   124 │   │   if not self.is_clone:                                                                                                                                              │
│   125 │   │   │   self.logger.add_log_file_extra(self.tools.android_sdk.list_packages)                                                                                           │
│   126                                                                                                                                                                            │
│                                                                                                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮                                                                                 │
│ │ self = <briefcase.platforms.android.gradle.GradleCreateCommand object at 0x0000024864293CA0> │                                                                                 │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯                                                                                 │
│                                                                                                                                                                                  │
│ D:\beeware-venv\lib\site-packages\briefcase\integrations\android_sdk.py:203 in verify                                              │
│                                                                                                                                                                                  │
│    200 │   │   │   if sdk.exists():                                                                                                                                              │
│    201 │   │   │   │   # NOTE: For now, all known versions of the cmdline-tools are compatible.                                                                                  │
│    202 │   │   │   │   # If/when that ever changes, do a verification check here.                                                                                                │
│ ❱  203 │   │   │   │   sdk.verify_license()                                                                                                                                      │
│    204 │   │   │   else:                                                                                                                                                         │
│    205 │   │   │   │   # The legacy SDK Tools exist. Delete them.                                                                                                                │
│    206 │   │   │   │   if (sdk_root_path / "tools").exists():                                                                                                                    │
│                                                                                                                                                                                  │
│ ╭─────────────────────────────────────────────────── locals ───────────────────────────────────────────────────╮                                                                 │
│ │           cls = <class 'briefcase.integrations.android_sdk.AndroidSDK'>                                      │                                                                 │
│ │       install = True                                                                                         │                                                                 │
│ │           sdk = <briefcase.integrations.android_sdk.AndroidSDK object at 0x000002486638B7C0>                 │                                                                 │
│ │      sdk_root = None                                                                                         │                                                                 │
│ │ sdk_root_path = WindowsPath('C:/Users/ccinsz/AppData/Local/BeeWare/briefcase/Cache/tools/android_sdk') │                                                                 │
│ │         tools = <briefcase.integrations.base.ToolCache object at 0x0000024865A65D60>                         │                                                                 │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────╯                                                                 │
│                                                                                                                                                                                  │
│ D:\beeware-venv\lib\site-packages\briefcase\integrations\android_sdk.py:398 in verify_license                                      │
│                                                                                                                                                                                  │
│    395 │   │   │   ) from e                                                                                                                                                      │
│    396 │   │                                                                                                                                                                     │
│    397 │   │   if not license_path.exists():                                                                                                                                     │
│ ❱  398 │   │   │   raise BriefcaseCommandError(                                                                                                                                  │
│    399 │   │   │   │   """\                                                                                                                                                      │
│    400 You did not accept the Android SDK licenses. Please re-run the briefcase command                                                                                          │
│    401 and accept the Android SDK license when prompted. You may need an Internet                                                                                                │
│                                                                                                                                                                                  │
│ ╭───────────────────────────────────────────────────────────────── locals ─────────────────────────────────────────────────────────────────╮                                     │
│ │ license_path = WindowsPath('C:/Users/ccinsz/AppData/Local/BeeWare/briefcase/Cache/tools/android_sdk/licenses/android-sdk-license') │                                     │
│ │         self = <briefcase.integrations.android_sdk.AndroidSDK object at 0x000002486638B7C0>                                              │                                     │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯                                     │
╰──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────╯
BriefcaseCommandError: You did not accept the Android SDK licenses. Please re-run the briefcase command
and accept the Android SDK license when prompted. You may need an Internet
connection.
```

**Reply by freakboy3742** [reply]
Firstly - I suspect this *isn't* the entire log file - there's an entire Environment Variables section missing.

Secondly - The log message tells you the source of the problem many times over - the Android licenses haven't been accepted. I don't know how you got into this state - the most likely answer is that you didn't accept the licenses when prompted to do so during an earlier run; but it's also possible that something about your local network configuration has prevented the license acceptance process from working.

You should be able to reset the Android tools by deleting the folder named `C:\Users\ccinsz\AppData\Local\BeeWare\briefcase\Cache\tools\android_sdk\licenses` and re-running Briefcase. This should prompt you to accept the licenses again. If that doesn't work, delete the entire `C:\Users\ccinsz\AppData\Local\BeeWare\briefcase\Cache\tools\android_sdk` folder and try again.

**Reply by ccinsz** [reply]
Yes, I try to delete the entire folder, but it does not work. When I use "briefcase create android", it shows like that:
The Android tools provided by Google have license terms that you must accept
before you may use those tools.
All SDK package licenses accepted.======] 100% Computing updates...

You did not accept the Android SDK licenses. Please re-run the briefcase command
and accept the Android SDK license when prompted. You may need an Internet
connection.

If my local network configuration has prevented, every version of the briefcase  should not work. But the briefcase==0.3.8 can work. This is my doubt.

**Reply by mhsmith** [reply]
> I try to delete the entire folder, 

Which folder exactly?

> but it does not work

Do you mean you can't delete it, or you successfully deleted it but it didn't solve the problem?

If it's the second one, please post a list of the files in the `licenses` folder.

Finally, please post the missing environment variables section of the log file, as @freakboy3742 mentioned above.

**Reply by ccinsz** [reply]
The second one. I delete the .briefcase file, and run  "briefcase create android", but it can't download the licenses folder. 
Where can I get the licenses folder?

The missing environment variables are as follows,

Environment Variables:
    ALLUSERSPROFILE=C:\ProgramData
    APPDATA=C:\Users\ccinsz\AppData\Roaming
    CHOCOLATEYINSTALL=C:\ProgramData\chocolatey
    COMMONPROGRAMFILES=C:\Program Files\Common Files
    COMMONPROGRAMFILES(X86)=C:\Program Files (x86)\Common Files
    COMMONPROGRAMW6432=C:\Program Files\Common Files
    COMPUTERNAME=NTDP-8TK0TK3
    COMSPEC=C:\Windows\system32\cmd.exe
    DRIVERDATA=C:\Windows\System32\Drivers\DriverData
    HOMEDRIVE=C:
    HOMEPATH=\Users\ccinsz
    INTELLIJ IDEA COMMUNITY EDITION=D:\software\IntelliJ IDEA Community Edition 2022.2.3\bin;
    IS_ETHERNET_SELECTED=false
    IS_WWAN_SELECTED=true
    LOCALAPPDATA=C:\Users\ccinsz\AppData\Local
    LOGONSERVER=\\TW-AD01
    NODE_PATH=C:\Program Files\nodejs
    NUMBER_OF_PROCESSORS=16
    ONEDRIVE=C:\Users\ccinsz\OneDrive
    OS=Windows_NT
    PATH=D:\beeware-venv\Scripts;C:\Program Files\Common Files\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;D:\software\xshell;D:\software\SDK\platform-tools;C:\Program Files\dotnet;D:\software\SDK\build-tools\33.0.0;C:\Program Files\nodejs;D:\software\gcc\mingw64\bin;D:\software\Git\cmd;D:\software\python\Scripts\;D:\software\python\;C:\Users\ccinsz\AppData\Local\Microsoft\WindowsApps;D:\software\PyCharm Community Edition 2021.3.2\bin;D:\software\IntelliJ IDEA Community Edition 2022.2.3\bin
    PATHEXT=.COM;.EXE;.BAT;.CMD;.VBS;.VBE;.JS;.JSE;.WSF;.WSH;.MSC;.CPL
    PROCESSOR_ARCHITECTURE=AMD64
    PROCESSOR_IDENTIFIER=Intel64 Family 6 Model 167 Stepping 1, GenuineIntel
    PROCESSOR_LEVEL=6
    PROCESSOR_REVISION=a701
    PROGRAMDATA=C:\ProgramData
    PROGRAMFILES=C:\Program Files
    PROGRAMFILES(X86)=C:\Program Files (x86)
    PROGRAMW6432=C:\Program Files
    PROMPT=(beeware-venv) $P$G
    PSMODULEPATH=C:\Users\ccinsz\Documents\WindowsPowerShell\Modules;C:\Program Files\WindowsPowerShell\Modules;C:\Windows\system32\WindowsPowerShell\v1.0\Modules
    PUBLIC=C:\Users\Public
    PYCHARM COMMUNITY EDITION=D:\software\PyCharm Community Edition 2021.3.2\bin;
    SYSTEMDRIVE=C:
    SYSTEMROOT=C:\Windows
    TEMP=C:\Users\AppData\Local\Temp
    TMP=C:\Users\AppData\Local\Temp
    USERDNSDOMAIN=CCINSZ.LOCAL
    USERDOMAIN=CCINSZ
    USERDOMAIN_ROAMINGPROFILE=CCINSZ
    USERNAME=ccinsz
    USERPROFILE=C:\Users\ccinsz
    VIRTUAL_ENV=D:\beeware-venv
    WINDIR=C:\Windows
    WSLENV=WT_SESSION::WT_PROFILE_ID
    WT_PROFILE_ID={72c54bbd-c2c6-5271-94e7-009a87ff44bf}
    WT_SESSION=d24bc97e-fce9-4845-8b56-65cc04bec772
    ZES_ENABLE_SYSMAN=1
    _OLD_VIRTUAL_PATH=C:\Program Files\Common Files\Oracle\Java\javapath;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0;D:\software\xshell;D:\software\SDK\platform-tools;C:\Program Files\dotnet;D:\software\SDK\build-tools\33.0.0;C:\Program Files\nodejs;D:\software\gcc\mingw64\bin;D:\software\Git\cmd;D:\software\python\Scripts\;D:\software\python\;C:\Users\ccinsz\AppData\Local\Microsoft\WindowsApps;D:\software\PyCharm Community Edition 2021.3.2\bin;D:\software\IntelliJ IDEA Community Edition 2022.2.3\bin
    _OLD_VIRTUAL_PROMPT=$P$G

**Reply by freakboy3742** [reply]
> The second one. I delete the .briefcase file, and run "briefcase create android", but it can't download the licenses folder. 

The `.briefcase` folder was used by old version of Briefcase. It isn't being used any more. Deleting this folder will recover disk space, but won't alter the operation of Briefcase 0.3.14.

> Where can I get the licenses folder?

I gave you very specific instructions for what to do, and they didn't involve deleting a `.briefcase` folder. Please follow those instructions.

**Reply by ccinsz** [reply]
I delete the entire C:\Users\ccinsz\AppData\Local\BeeWare folder and try again. And it has the license terms, I accept all, but it also raises the same problem.
![image](https://github.com/beeware/briefcase/assets/134466342/dc7de3c9-0110-42a6-88ce-36242ba3c261)

**Reply by mhsmith** [reply]
Does `BeeWare\briefcase\Cache\tools\android_sdk\licenses` exist? If so, please post a list of the files it contains.

**Reply by ccinsz** [reply]
The problem of the licenses is solved. Thank you very much for your help @freakboy3742 @mhsmith.

**Reply by mhsmith** [reply]
Your deleted screenshot showed that all the license files were present, but with `.IPGSD` extensions, which I can't find any reference to on Google. 

Please tell us how you solved this, so we can help other users with the same problem.

**Reply by ccinsz** [reply]
Yes, my computer will encrypt the files, I write the copy to delete the files .IPGSD extensions of the files, and solved the problem. I don't think others will have this problem.
